tiledb/sm/storage_manager/context.cc

```diff
@@ -45,62 +45,35 @@ namespace sm {
 /*   CONSTRUCTORS & DESTRUCTORS   */
 /* ****************************** */
 
-Context::Context()
-    : last_error_(Status::Ok())
-    , storage_manager_(nullptr)
-    , stats_(make_shared<stats::Stats>(HERE(), "Context"))
-    , logger_(make_shared<Logger>(
-          HERE(), "Context: " + std::to_string(++logger_id_))) {
+Context::Context(const Config& config) try
+    : last_error_(Status::Ok()),
+      compute_tp_(get_compute_thread_count(config)),
+      io_tp_(get_io_thread_count(config)),
+      stats_(make_shared<stats::Stats>(HERE(), "Context")),
+      logger_(make_shared<Logger>(
+          HERE(), "Context: " + std::to_string(++logger_id_))),
+      storage_manager_(nullptr) {
+  if (!init(config).ok()) {
+    logger_->status(
+        Status_ContextError("Failed to initialize Context in constructor."));
+    cleanup_storage_manager();
+    throw std::logic_error("Could not initialize Context in constructor");
+  }
+} catch (...) {
+  // any exceptions in member initializer list will have happenened before
+  // storage_manager is initialized, so we don't need to clean up anything here.
+  // So just rethrow;
+  throw;
```

Comment:
Unless we're going to log something here, this try-statement has no different behavior than leaving it out entirely would.

The comment here about initialization order is better suited to documentation about the member variable itself. If there's a requirement that some member variable be initialized last, it should be local to that variable. Really, though, "last" isn't the most descriptive way of stating it. The documentation should state clearly what other member variables the storage manager depends upon, and state the requirement that it be declared after them.

Reply:
I prefer to have the comment about initialization order near to the location where the initialization is happening and the ordering is shown -- the comment is also to reassure the reader about exception-safety.

I think I left the `try-catch` there as a placeholder as I meant to ask about what the appropriate log message would be.

Reply:
(Adding a log message.)

Reply:
>comment about initialization order

In practice for this code base, declaration order is initialization order. If not, there are warnings that escalate into errors and it doesn't pass CI.

The reason to put initialization order comments near the member variable declarations is that's where code maintenance must happen. If a member variable has an initializer, changes in its initialization order may not show up in the constructor. I don't mind making the comment in both places, to be sure.

A comment near the constructor say what is happening, but a comment near the declaration acts as a requirement that says why it's happening. If I had to pick one, I'd go for the requirement. I'm OK with having both.

Reply:
Done.